src/plugins/intel_cpu/src/nodes/segment_max.cpp

```diff
@@ -0,0 +1,153 @@
+    : Node(op, context, NgraphShapeInferFactory(op)) {
+    std::string errorMessage;
+    if (!isSupportedOperation(op, errorMessage)) {
+        OPENVINO_THROW_NOT_IMPLEMENTED(errorMessage);
+    }
+    fillMode = ov::as_type_ptr<const ov::op::v16::SegmentMax>(op)->get_fill_mode();
+}
+
+bool SegmentMax::isSupportedOperation(const std::shared_ptr<const ov::Node>& op, std::string& errorMessage) noexcept {
+    try {
+        if (!ov::is_type<ov::op::v16::SegmentMax>(op)) {
+            errorMessage = "Only opset16 SegmentMax operation is supported";
+            return false;
+        }
+    } catch (...) {
+        return false;
+    }
+    return true;
+}
+
+void SegmentMax::getSupportedDescriptors() {
+    // Validation is already done in the ov::opset16::SegmentMax
+}
+
+void SegmentMax::initSupportedPrimitiveDescriptors() {
+    if (!supportedPrimitiveDescriptors.empty()) {
+        return;
+    }
+    ov::element::Type dataPrecision = getOriginalInputPrecisionAtPort(0);
+    if (getOriginalInputsNumber() == 2) {
+        addSupportedPrimDesc({{LayoutType::ncsp, dataPrecision}, {LayoutType::ncsp, ov::element::i32}},
+                             {{LayoutType::ncsp, dataPrecision}},
+                             impl_desc_type::ref);
+    } else {
+        addSupportedPrimDesc({{LayoutType::ncsp, dataPrecision},
+                              {LayoutType::ncsp, ov::element::i32},
+                              {LayoutType::ncsp, ov::element::i32}},
+                             {{LayoutType::ncsp, dataPrecision}},
+                             impl_desc_type::ref);
+    }
+}
+
+bool SegmentMax::created() const {
+    return getType() == Type::SegmentMax;
+}
+
+bool SegmentMax::needPrepareParams() const {
+    return false;
+}
+
+void SegmentMax::executeDynamicImpl(const dnnl::stream& strm) {
+    execute(strm);
+
+    // Update lastSegmentIds
+    const auto* srcSegmentIds = getSrcDataAtPortAs<const int32_t>(1);
+    lastSegmentIds.assign(srcSegmentIds, srcSegmentIds + getSrcMemoryAtPort(1)->getSize());
+
+    // Update lastNumSegments
+    if (getOriginalInputsNumber() == 3) {
+        const auto* numSegmentsValue = getSrcDataAtPortAs<const int32_t>(2);
+        if (lastNumSegments.empty()) {
+            lastNumSegments.push_back(*numSegmentsValue);
+        } else {
+            lastNumSegments[0] = *numSegmentsValue;
+        }
+    }
+}
+
+bool SegmentMax::needShapeInfer() const {
+    if (inputShapesModified()) {
+        return true;
+    }
+    if (lastSegmentIds.empty()) {
+        return true;
+    }
+
+    // Check if segmentIds has changed
+    if (lastSegmentIds.size() != getSrcMemoryAtPort(1)->getSize()) {
+        return true;
+    }
+    const auto* segmentIds = getSrcDataAtPortAs<const int32_t>(1);
+    for (size_t i = 0; i < lastSegmentIds.size(); i++) {
+        if (lastSegmentIds[i] != segmentIds[i]) {
+            return true;
+        }
+    }
```

Comment:
It the Tensor with SegmentIds can be really big, there is a chance that recalculation of shape can be less consuming than comparison of each segment. But I would consider it as "nice to check" for this PR.